resources/user.rb

```diff
@@ -0,0 +1,161 @@
+property :first_name, String, default: ''
+property :last_name, String, default: ''
+property :email, String, default: ''
+property :roles, Array, default: []
+property :api_endpoint, String, identity: true, default: lazy { node['nexus3']['api']['endpoint'] }
+property :api_username, String, identity: true, default: lazy { node['nexus3']['api']['username'] }
+property :api_password, String, identity: true, sensitive: true, default: lazy { node['nexus3']['api']['password'] }
+
+load_current_value do |desired|
+  apiclient = ::Nexus3::Api.new(api_endpoint, api_username, api_password)
+
+  begin
+    res = apiclient.run_script('get_user', username)
+    current_value_does_not_exist! if res == 'null'
+    config = JSON.parse(res)
+    ::Chef::Log.warn "User config is #{config}"
+    first_name config['first_name']
+    last_name config['last_name']
+    email config['email']
+    roles config['roles']
+
+    # Check if we need to change the password.
+    apiwithuser = ::Nexus3::Api.new(api_endpoint, username, desired.password)
+    begin
+      apiwithuser.request(:get, '/service/metrics/ping')
+    rescue ::Nexus3::ApiError
+      password desired.password
+    else
+      password 'Supercalifragilisticexpialidocious-that-does-not-exist-so-maybe-the-resource-will-need-to-converge'
+    end
+  rescue LoadError, ::Nexus3::ApiError => e
+    ::Chef::Log.warn "A '#{e.class}' occurred: #{e.message}"
+    current_value_does_not_exist!
+  end
+end
+
+action :create do
+  init
+
+  converge_if_changed do
+    nexus3_api "upsert_user #{new_resource.username}" do
+      script_name 'upsert_user'
+      args username: new_resource.username,
+           password: new_resource.password,
+           first_name: new_resource.first_name,
+           last_name: new_resource.last_name,
+           email: new_resource.email,
+           roles: new_resource.roles
+
+      action %i(create run)
+      endpoint new_resource.api_endpoint
+      username new_resource.api_username
+      password new_resource.api_password
+
+      content <<-EOS
+import org.sonatype.nexus.security.role.NoSuchRoleException;
+import org.sonatype.nexus.security.role.RoleIdentifier;
+import org.sonatype.nexus.security.user.User;
+import org.sonatype.nexus.security.user.UserManager;
+import org.sonatype.nexus.security.user.UserNotFoundException;
+import org.sonatype.nexus.security.SecuritySystem;
+
+import groovy.json.JsonSlurper;
+
+def params = new JsonSlurper().parseText(args)
+
+authManager = security.getSecuritySystem().getAuthorizationManager(UserManager.DEFAULT_SOURCE);
+// Create a list of available from the list of passed roles. May end up empty if no role exists.
+Set<RoleIdentifier> roleList = new HashSet<RoleIdentifier>();
+params.roles.each { role ->
+    try {
+        def newRole = authManager.getRole(role);
+        roleList.add(newRole);
+    } catch (NoSuchRoleException e) {
+        log.info("No such role: ${role}, trying to set for ${params.username} from Chef");
```

Comment:
If the role doesn't exist you onl log an Info message? I would go for a warning

Reply:
Done